liquibase-core/src/main/java/liquibase/lockservice/ProlongingLockService.java

```diff
@@ -0,0 +1,879 @@
+import liquibase.statement.core.UnlockDatabaseChangeLogStatement;
+import liquibase.structure.core.Table;
+
+/**
+ * Prolongs the current lock every 30 seconds.
+ * <p>
+ * Locks older than 100 seconds are considered stale and will get removed.
+ * <p>
+ * Since locks get actively prolonged, do use a very high lock wait time (1 day).
+ * <p>
+ * Must <b>not</b> be used in junction with the {@link StandardLockService}!
+ * <p>
+ * Note: This is in alpha state. We do not use the time of the database server right now,
+ * but a local date, so this might not work properly when several services with very
+ * un-synchronised locks connect to the same database.
+ */
+public class ProlongingLockService implements LockService {
+
+    private static final int LOW_PRIO = -1;
+    private static final int PRIO_HIGHER_THAN_STANDARD_LOCK_SERVICE = 2;
+
+    // Only prolong your own lock, hence use a unique id for each service
+    // accessing the database
+    private static final String LOCKED_BY_ID = UUID.randomUUID().toString();
+
+    protected Database database;
+
+    /**
+     * This is a local timestamp. We record when our lock would be stale.
+     * <p>
+     * On the server, however, we use current_timestamp instead.
+     * <p>
+     * So these two values could be different, but as long as we compare
+     * values on the server with each other, and local Dates with each other,
+     * it's not a problem.
+     * <p>
+     * Treat this as the local, cached version of the value that is written
+     * in LOCKEXPIRES in the database.
+     */
+    private volatile Date lockExpires = null;
+
+    private volatile ScheduledFuture<?> lockProlonger = null;
+
+    /**
+     * How long shall we wait for a lock before we give up?
+     * <p>
+     * Unit: minutes
+     */
+    private Long changeLogLockWaitTime;
+
+    /**
+     * How often shall we recheck
+     * <p>
+     * Should be &gt; {@link #changeLogLockProlongingRateInSeconds},
+     * <p>
+     * Units: seconds
+     */
+    private Long changeLogLockRecheckTime;
+
+    /**
+     * How often do we want to prolong the lock?
+     * <p>
+     * Unit: seconds
+     */
+    private Long changeLogLockProlongingRateInSeconds;
+
+    /**
+     * After what time shall we remove a lock?
+     * <p>
+     * This must be larger than {@link #changeLogLockProlongingRateInSeconds}. Ideally add 40
+     * seconds to that.
+     * <p>
+     * If this is just a little larger than {@link #changeLogLockProlongingRateInSeconds} you
+     * might end up
+     * removing locks another (active) service is just about to prolong.
+     * <p>
+     * Unit: seconds
+     */
+    private Long staleChangeLogLockRemovalTimeInSeconds;
+
+    private Boolean hasDatabaseChangeLogLockTable;
+    private boolean isDatabaseChangeLogLockTableInitialized;
+    private ObjectQuotingStrategy quotingStrategy;
+
+    private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1);
```

Comment:
I feel like there are now thread-safety issues in this class that have not been fully thought through.
I think that there need to be a few more synchronised blocks where the main thread interacts with the prolonging thread, but I have not analysed it in detail.

Reply:
you think about memory effects? For this I have added volatile to some fields. Or you mean execution order?